docs/spec/fastsync/fastsync.md

```diff
@@ -0,0 +1,1163 @@
+- Expected precondition
+    - *peerHeights(addr) <= ht*
+- Expected postcondition
+    - *peerHeights(addr) = ht*
+	- *TargetHeight* is updated
+- Error condition
+    - if precondition is violated: *addr* not in *peerIDs* (that is,
+      *addr* is removed from *peerIDs*)
+- Timeout condition
+    - if `OnStatusResponse(addr, ht)` was not invoked within *2 Delta* after
+	`Status(addr)` was called:  *addr* not in *peerIDs*
+----
+
+
+```go
+func CreateRequest
+```
+- Expected precondition
+    - *height < TargetHeight*
+	- *peerIDs* nonempty
+- Expected postcondition
+    - Function `Block` is called remotely at a peer *addr* in peerIDs 
+	  for a missing height *h*  
+	  *Remark:* different implementations may have different
+      strategies to balance the load over the peers
+    - *pendingblocks(h) = addr*
+----
+
+
+```go
+func OnBlockResponse(addr Address, b Block)
+```
+- Comment
+    - if after adding block *b*, blocks of heights *height* and
+      *height + 1* are in *blockstore*, then `Execute` is called
+- Expected precondition
+    - *pendingblocks(b.Height) = addr*
+	- *b* satisfies basic soundness  
+- Expected postcondition
+    - if function `Execute` has been executed without error or was not
+      executed:
+        - *receivedBlocks(b.Height) = addr*
+        - *blockstore(b.Height) = b*
+        - *peerTimeStamp[addr]* is set to a time between invocation and
+          return of the function.
+        - *peerRate[addr]* is updated according to size of received
+          block
+- Error condition
+    - if precondition is violated: *addr* not in *peerIDs*; reset
+	*pendingblocks(b.Height)* to nil;
+- Timeout condition
+    - if `OnBlockResponse(addr, b)` was not invoked within *2 Delta* after
+	`Block(addr,h)` was called for *b.Height = h*: *addr* not in *peerIDs*
+----
+
+
+
+```go
+func Execute()
+```
+- Comments
+    - none
+- Expected precondition
+	- application state is the one of the blockchain at height
+      *height - 1*
+	- **[FS-V2-Verif]** for any two blocks *a* and *b* from
+	*receivedBlocks*: if
+	  *a.Height + 1 = b.Height* then *VerifyCommit (a,b.Commit) = true*
+- Expected postcondition
+    - Any two blocks *a* and *b* violating [FS-V2-Verif]:
+	  *a* and *b* not in *blockstore*; nodes with Address 
+	  receivedBlocks(a.Height) and receivedBlocks(b.Height) not in peerIDs
+	- height is updated height of complete prefix that matches the blockchain
+	- state is the one of the blockchain at height *height - 1*
+- Error condition
+    - none
+----
+
+## Analysis of Fastsync V2
+
+####  **[FS-ISSUE-KILL]**:
+If two blocks are not matching [FS-V2-Verif], `Execute` dismisses both
+blocks and removes the peers that provided these blocks from
+*peerIDs*. If block *a* was correct and provided by a correct peer *p*,
+and block b was faulty and provided by a faulty peer, the protocol
+- removes the correct peer *p*, although it might be useful to
+  download blocks from it in the future
```

Comment:
Maybe it's better to link it with failed invariant (we don't download all blocks from correct peers).